go/vt/sqlparser/ast.go

```diff
@@ -33,6 +34,36 @@ import (
 	vtrpcpb "vitess.io/vitess/go/vt/proto/vtrpc"
 )
 
+// parserPool is a pool for parser objects.
+var parserPool = sync.Pool{}
+
+// zeroParser is a zero-initialized parser to help reinitialize the parser for pooling.
+var zeroParser = *(yyNewParser().(*yyParserImpl))
+
+// yyParsePooled is a wrapper around yyParse that pools the parser objects. There isn't a
+// particularly good reason to use yyParse directly, since it immediately discards its parser.  What
+// would be ideal down the line is to actually pool the stacks themselves rather than the parser
+// objects, as per https://github.com/cznic/goyacc/blob/master/main.go. However, absent an upstream
+// change to goyacc, this is the next best option.
+func yyParsePooled(yylex yyLexer) int {
+	// Being very particular about using the base type and not an interface type b/c we depend on
+	// the implementation to know how to reinitialize the parser.
+	var parser *yyParserImpl
+
+	i := parserPool.Get()
+	if i != nil {
+		parser = i.(*yyParserImpl)
+	} else {
+		parser = yyNewParser().(*yyParserImpl)
+	}
+
+	defer func() {
+		*parser = zeroParser
```

Comment:
unsurprisingly, this turns out to be a problem:

```--- FAIL: TestValid (0.00s)
panic: interface conversion: interface is nil, not sqlparser.SQLNode [recovered]
        panic: interface conversion: interface is nil, not sqlparser.SQLNode

goroutine 77 [running]:
testing.tRunner.func1(0xc4206823c0)
        /usr/local/go/src/testing/testing.go:742 +0x29d
panic(0x662e60, 0xc42070c180)
        /usr/local/go/src/runtime/panic.go:502 +0x229
vitess.io/vitess/go/vt/sqlparser.(*TrackedBuffer).Myprintf(0xc420164900, 0x6b7bd9, 0x9, 0xc42005dd08, 0x1, 0x1)
        /home/tahara/vitess/src/vitess.io/vitess/go/vt/sqlparser/tracked_buffer.go:97 +0x3ba
vitess.io/vitess/go/vt/sqlparser.(*Show).Format(0xc4200824b0, 0xc420164900)
        /home/tahara/vitess/src/vitess.io/vitess/go/vt/sqlparser/ast.go:1498 +0x1a3
vitess.io/vitess/go/vt/sqlparser.(*TrackedBuffer).Myprintf(0xc420164900, 0x6b60ed, 0x2, 0xc42005de70, 0x1, 0x1)
        /home/tahara/vitess/src/vitess.io/vitess/go/vt/sqlparser/tracked_buffer.go:99 +0x3ff
vitess.io/vitess/go/vt/sqlparser.String(0x7fd7c9a1ee68, 0xc4200824b0, 0xc4200824b0, 0x7fd7c9a1ee68)
        /home/tahara/vitess/src/vitess.io/vitess/go/vt/sqlparser/ast.go:249 +0xe4
vitess.io/vitess/go/vt/sqlparser.TestValid(0xc4206823c0)
        /home/tahara/vitess/src/vitess.io/vitess/go/vt/sqlparser/parse_test.go:1347 +0x20a
testing.tRunner(0xc4206823c0, 0x6d2920)
        /usr/local/go/src/testing/testing.go:777 +0xd0
created by testing.(*T).Run
        /usr/local/go/src/testing/testing.go:824 +0x2e0
exit status 2
FAIL    vitess.io/vitess/go/vt/sqlparser        0.040s
```

Reply:
Fixed. there was a dangling reference from ShowCollationFilterOpt, which is resolved in this diff